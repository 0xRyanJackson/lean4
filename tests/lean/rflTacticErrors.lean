
/-!
This file tests the `rfl` tactic:
 * Extensibility
 * Error messages
 * Effect of `with_reducible`
-/

-- Remove the following once `apply_rfl` and `rfl` are merged
-- Until then: Note the horrible error message!
example : false = true   := by rfl

-- Setup

inductive P : α → α → Prop where | refl : P a a
attribute [refl] P.refl

-- Defeq to relation `P` at reducible transparency
abbrev Q : α → α → Prop := P
-- Defeq to relation `P` at default transparency
def Q' : α → α → Prop := P

-- No refl attribute
inductive R : α → α → Prop where | refl : R a a

-- Syntactic equal

example : true = true   := by apply_rfl
example : HEq true true := by apply_rfl
example : True ↔ True   := by apply_rfl
example : P true true   := by apply_rfl
example : Q true true   := by apply_rfl
example : Q' true true  := by apply_rfl -- Error
example : R true true   := by apply_rfl -- Error

example : true = true   := by with_reducible apply_rfl
example : HEq true true := by with_reducible apply_rfl
example : True ↔ True   := by with_reducible apply_rfl
example : P true true   := by with_reducible apply_rfl
example : Q true true   := by with_reducible apply_rfl
example : Q' true true  := by with_reducible apply_rfl -- Error
example : R true true   := by with_reducible apply_rfl -- Error

-- Reducibly equal

abbrev true' := true
abbrev True' := True

example : true' = true   := by apply_rfl
example : HEq true' true := by apply_rfl
example : True' ↔ True   := by apply_rfl
example : P true' true   := by apply_rfl
example : Q true' true   := by apply_rfl
example : Q' true' true  := by apply_rfl -- Error
example : R true' true   := by apply_rfl -- Error

example : true' = true   := by with_reducible apply_rfl
example : HEq true' true := by with_reducible apply_rfl
example : True' ↔ True   := by with_reducible apply_rfl
example : P true' true   := by with_reducible apply_rfl
example : Q true' true   := by with_reducible apply_rfl -- NB: No error!
example : Q' true' true  := by with_reducible apply_rfl -- Error
example : R true' true   := by with_reducible apply_rfl -- Error

-- Equal at default transparency only

def true'' := true
def True'' := True

example : true'' = true   := by apply_rfl
example : HEq true'' true := by apply_rfl
example : True'' ↔ True   := by apply_rfl
example : P true'' true   := by apply_rfl
example : Q true'' true   := by apply_rfl
example : Q' true'' true  := by apply_rfl -- Error
example : R true'' true   := by apply_rfl -- Error

example : true'' = true   := by with_reducible apply_rfl -- Error
example : HEq true'' true := by with_reducible apply_rfl -- Error
example : True'' ↔ True   := by with_reducible apply_rfl -- Error
example : P true'' true   := by with_reducible apply_rfl -- Error
example : Q true'' true   := by with_reducible apply_rfl -- Error
example : Q' true'' true  := by with_reducible apply_rfl -- Error
example : R true'' true   := by with_reducible apply_rfl -- Error

-- Unequal
example : false = true   := by apply_rfl -- Error
example : HEq false true := by apply_rfl -- Error
example : False ↔ True   := by apply_rfl -- Error
example : P false true   := by apply_rfl -- Error
example : Q false true   := by apply_rfl -- Error
example : Q' false true  := by apply_rfl -- Error
example : R false true   := by apply_rfl -- Error

example : false = true   := by with_reducible apply_rfl -- Error
example : HEq false true := by with_reducible apply_rfl -- Error
example : False ↔ True   := by with_reducible apply_rfl -- Error
example : P false true   := by with_reducible apply_rfl -- Error
example : Q false true   := by with_reducible apply_rfl -- Error
example : Q' false true  := by with_reducible apply_rfl -- Error
example : R false true   := by with_reducible apply_rfl -- Error

-- Inheterogeneous unequal

example : HEq true 1 := by apply_rfl -- Error
example : HEq true 1 := by with_reducible apply_rfl -- Error

-- Rfl lemma with side condition:
-- Error message should show left-over goal

inductive S : Bool → Bool → Prop where | refl : a = true → S a a
attribute [refl] S.refl
example : S true false  := by apply_rfl -- Error
example : S true false  := by with_reducible apply_rfl -- Error
example : S true true   := by apply_rfl -- Error (left-over goal)
example : S true true   := by with_reducible apply_rfl -- Error (left-over goal)
example : S false false := by apply_rfl -- Error (left-over goal)
example : S false false := by with_reducible apply_rfl -- Error (left-over goal)
