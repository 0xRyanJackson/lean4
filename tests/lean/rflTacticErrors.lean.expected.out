rflTacticErrors.lean:11:31-11:34: error: type mismatch
  Iff.rfl
has type
  ?m ↔ ?m : Prop
but is expected to have type
  false = true : Prop
rflTacticErrors.lean:33:30-33:39: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:34:30-34:39: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:41:45-41:54: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:42:45-42:54: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:54:31-54:40: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:55:31-55:40: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:62:46-62:55: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:63:46-63:55: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:75:32-75:41: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:76:32-76:41: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:78:47-78:56: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ true'' = true
rflTacticErrors.lean:79:47-79:56: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq true'' true
⊢ HEq true'' true
rflTacticErrors.lean:80:47-80:56: error: tactic 'rfl' failed, The lhs
  True''
is not definitionally equal to rhs
  True
⊢ True'' ↔ True
rflTacticErrors.lean:81:47-81:56: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ P true'' true
rflTacticErrors.lean:82:47-82:56: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ Q true'' true
rflTacticErrors.lean:83:47-83:56: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ Q' true'' true
rflTacticErrors.lean:84:47-84:56: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ R true'' true
rflTacticErrors.lean:87:31-87:40: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ false = true
rflTacticErrors.lean:88:31-88:40: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq false true
⊢ HEq false true
rflTacticErrors.lean:89:31-89:40: error: tactic 'rfl' failed, The lhs
  False
is not definitionally equal to rhs
  True
⊢ False ↔ True
rflTacticErrors.lean:90:31-90:40: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ P false true
rflTacticErrors.lean:91:31-91:40: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q false true
rflTacticErrors.lean:92:31-92:40: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q' false true
rflTacticErrors.lean:93:31-93:40: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ R false true
rflTacticErrors.lean:95:46-95:55: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ false = true
rflTacticErrors.lean:96:46-96:55: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq false true
⊢ HEq false true
rflTacticErrors.lean:97:46-97:55: error: tactic 'rfl' failed, The lhs
  False
is not definitionally equal to rhs
  True
⊢ False ↔ True
rflTacticErrors.lean:98:46-98:55: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ P false true
rflTacticErrors.lean:99:46-99:55: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q false true
rflTacticErrors.lean:100:46-100:55: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q' false true
rflTacticErrors.lean:101:46-101:55: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ R false true
rflTacticErrors.lean:105:27-105:36: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq true 1
⊢ HEq true 1
rflTacticErrors.lean:106:42-106:51: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq true 1
⊢ HEq true 1
rflTacticErrors.lean:113:30-113:39: error: tactic 'rfl' failed, The lhs
  true
is not definitionally equal to rhs
  false
⊢ S true false
rflTacticErrors.lean:114:45-114:54: error: tactic 'rfl' failed, The lhs
  true
is not definitionally equal to rhs
  false
⊢ S true false
rflTacticErrors.lean:115:30-115:39: error: unsolved goals
case a
⊢ true = true
rflTacticErrors.lean:116:45-116:54: error: unsolved goals
case a
⊢ true = true
rflTacticErrors.lean:117:30-117:39: error: unsolved goals
case a
⊢ false = true
rflTacticErrors.lean:118:45-118:54: error: unsolved goals
case a
⊢ false = true
